igvm/hypervisor.py

```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
         the guest OS"""
         return 'vda1'
 
-    def migrate_vm(
-        self, vm, target_hypervisor, offline, offline_transport, transaction,
-        no_shutdown,
+    def check_migrate_vm(
+        self, vm: VM, offline: bool, offline_transport: str,
+        disk_size: int = 0,
     ):
-        if offline_transport not in ['netcat', 'drbd']:
+        if offline_transport not in ['netcat', 'drbd', 'xfs']:
             raise StorageError(
                 'Unknown offline transport method {}!'
                 .format(offline_transport)
             )
+        disk_size = disk_size if 0 < disk_size else 0
+        if not (offline and offline_transport == 'xfs') and 0 < disk_size:
+            raise StorageError(
+                'disk_size can be applied only with offline transport xfs!'
+            )
+        allocated_space = vm.dataset_obj['disk_size_gib'] - vm.disk_free()
+        if 0 < disk_size < allocated_space:
+            raise StorageError(
+                'disk_size is lower than allocated space: {} < {}!'
+                .format(disk_size, allocated_space)
+            )
+
+    def migrate_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        disk_size: int = 0,
+    ) -> int:
+        self.check_migrate_vm(vm, offline, offline_transport, disk_size)
+        disk_size = disk_size if 0 < disk_size else 0
+        return disk_size or vm.dataset_obj['disk_size_gib']
+
+    def migrate_change_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        transaction, disk_size: int = 0,
+    ):
+        size = self.migrate_disk_size(vm, offline,
+                                      offline_transport, disk_size)
+        if size == vm.dataset_obj['disk_size_gib']:
+            return
+
+        old_size = vm.dataset_obj['disk_size_gib']
+        vm.dataset_obj['disk_size_gib'] = size
+        vm.dataset_obj.commit()
+
+        if transaction:
+            def restore_size():
+                vm.dataset_obj['disk_size_gib'] = old_size
+                vm.dataset_obj.commit()
+            transaction.on_rollback('reset_disk_size', restore_size)
+
+
+    def migrate_wait_shutdown(
+        self, vm: VM, no_shutdown: bool, transaction: Transaction,
+    ):
+        vm.set_state('maintenance', transaction=transaction)
+        if vm.is_running():
+            if no_shutdown:
```

Comment:
I was not expecting that - maybe `wait_for_shutdown` or `manual_shutdown` is a better name.

Reply:
The parameter is the CLI argument, so `no_shutdown` is coming from the parser. I'll rename the method though.

Reply:
Renamed to `wait_for_shutdown`

Reply:
We should not allow changing the disk size without offline in the first place then. Similar to this: 
-  https://github.com/innogames/igvm/blob/master/igvm/commands.py#L418

Making the user think he can migrate the VM and change the disk size online and then ask him to shutdown the VM is counter intuitive and does not fit our already existing handling of similar cases. We are better of telling him right away that this is not possible.

Reply:
I guess the last comment is related to some other discussion.

Nevertheless, it looks more reasonable than to just replace https://github.com/innogames/igvm/blob/master/igvm/commands.py#L418 with `check_migrate_parameters`

Reply:
And `check_migrate_parameters` are already included here https://github.com/innogames/igvm/pull/176/commits/0371e62cd780a9044b36498ac7971049410228fb#diff-9299a928e2b4f7d2b1d51a5dcdd44a442d587f35480a7f2dfe8349022180e2cdR397, so the migration won't start on the wrong parameters